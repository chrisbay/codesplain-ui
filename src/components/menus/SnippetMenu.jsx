import React, { Component, PropTypes } from 'react';
import { Menu } from 'material-ui';
import SnippetList from './SnippetList';
import OrgSnippetsMenu from './OrgSnippetsMenu';
import CustomPropTypes from '../../util/custom-prop-types';

const styles = {
  iconMenu: {
    cursor: 'pointer',
  },
  menuText: {
    cursor: 'pointer',
    color: 'white',
  },
};

class SnippetMenu extends Component {
  constructor(props) {
    super(props);
    this.state = {
      menuOpen: false,
      snippetText: false,
      cursorOnMenuText: false,
    };
    this.handleOnRequestChange = this.handleOnRequestChange.bind(this);
    this.handleSnippetSelected = this.handleSnippetSelected.bind(this);
    this.toggleMenu = this.toggleMenu.bind(this);
    this.handleMenuTextHover = this.handleMenuTextHover.bind(this);
    this.handleMenuTextNotHovered = this.handleMenuTextNotHovered.bind(this);
  }

  toggleMenu() {
    this.setState({ menuOpen: !this.state.menuOpen });
  }

  handleOnRequestChange(open) {
    this.setState({
      menuOpen: open,
    });
  }

  handleSnippetSelected(snippetOwner, snippetKey) {
    this.setState({
      menuOpen: false,
    });
    this.props.onSnippetSelected(snippetOwner, snippetKey);
  }

  handleMenuTextHover() {
    this.setState({ cursorOnMenuText: true });
  }

  handleMenuTextNotHovered() {
    this.setState({ cursorOnMenuText: false });
  }

  render() {
    const {
      username,
      userSnippets,
      orgSnippets,
    } = this.props;
    const {
      menuOpen,
      cursorOnMenuText,
    } = this.state;

    const hoverStyles = cursorOnMenuText ? { textDecoration: 'underline' } : {};
    const menuText = (
      <div
        onClick={this.toggleMenu}
        style={{ ...hoverStyles, ...styles.menuText }}
        onMouseOver={this.handleMenuTextHover}
        onMouseLeave={this.handleMenuTextNotHovered}
      >
        Snippets
      </div>
    );
    const menu = menuOpen ?
      (<Menu
        anchorOrigin={{ horizontal: 'right', vertical: 'top' }}
        value="Snippets"
        onItemTouchTap={this.handleOnItemTouchTap}
        onRequestChange={this.handleOnRequestChange}
        open={menuOpen}
        style={styles.iconMenu}
        targetOrigin={{ horizontal: 'right', vertical: 'top' }}
        useLayerForClickAway
      >
        <SnippetList
          onClick={this.handleSnippetSelected}
          primaryText="My Snippets"
          snippetOwner={username}
          snippetsList={userSnippets}
        />
        <OrgSnippetsMenu
          onClick={this.handleSnippetSelected}
          orgSnippets={orgSnippets}
        />
      </Menu>) : null;

    return (
      <span>
        {menuText}
        {menu}
      </span>
    );
  }
}

SnippetMenu.propTypes = {
  username: PropTypes.string.isRequired,
  userSnippets: CustomPropTypes.snippets.isRequired,
  orgSnippets: CustomPropTypes.orgSnippets.isRequired,
  onSnippetSelected: PropTypes.func.isRequired,
};

export default SnippetMenu;
