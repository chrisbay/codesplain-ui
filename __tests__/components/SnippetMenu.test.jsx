import React from 'react';
import { shallow } from 'enzyme';
import getMuiTheme from 'material-ui/styles/getMuiTheme';
import { shallowToJson } from 'enzyme-to-json';
import SnippetMenu from '../../src/components/menus/SnippetMenu';

const defaultProps = {
  username: 'foo',
  userSnippets: {},
  orgSnippets: {},
  onSnippetSelected: jest.fn(),
  onHoverBackground: '',
  onOpenSearchMenu: jest.fn(),
  borderBottomColor: '',
};

describe('<SnippetMenu />', () => {
  const muiTheme = getMuiTheme();
  const shallowWithContext = node => shallow(node, { context: { muiTheme } });

  describe('snapshot tests', () => {
    it('matches snapshot of when user is not logged in', () => {
      const wrapper = shallowWithContext(
        <SnippetMenu {...defaultProps} />,
      );
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('matches snapshot of when user is logged in', () => {
      const wrapper = shallowWithContext(
        <SnippetMenu {...defaultProps} />,
      );
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });
});
