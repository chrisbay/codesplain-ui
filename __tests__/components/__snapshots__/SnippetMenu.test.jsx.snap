exports[`<SnippetMenu /> snapshot tests matches snapshot of when user is logged in 1`] = `
<IconMenu
  anchorOrigin={
    Object {
      "horizontal": "left",
      "vertical": "bottom",
    }
  }
  animated={true}
  iconButtonElement={
    <span
      style={
        Object {
          "color": "white",
        }
      }>
      Snippets
    </span>
  }
  multiple={false}
  onItemTouchTap={[Function]}
  onKeyboardFocus={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onRequestChange={[Function]}
  onTouchTap={[Function]}
  open={false}
  style={
    Object {
      "alignItems": "center",
      "background": "transparent",
      "borderBottomColor": "transparent",
      "borderBottomWidth": "3px",
      "cursor": "pointer",
      "display": "inline-flex",
      "height": "100%",
      "paddingLeft": "10px",
      "paddingRight": "10px",
    }
  }
  targetOrigin={
    Object {
      "horizontal": "left",
      "vertical": "top",
    }
  }
  touchTapCloseDelay={200}
  useLayerForClickAway={true}>
  <SnippetList
    onClick={[Function]}
    primaryText="My Snippets"
    snippetOwner="foo"
    snippetsList={Object {}} />
  <OrgSnippetsMenu
    onClick={[Function]}
    orgSnippets={Object {}} />
  <MenuItem
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    checked={false}
    desktop={false}
    disabled={false}
    focusState="none"
    insetChildren={false}
    onTouchTap={[Function]}
    primaryText="Search" />
</IconMenu>
`;

exports[`<SnippetMenu /> snapshot tests matches snapshot of when user is not logged in 1`] = `
<IconMenu
  anchorOrigin={
    Object {
      "horizontal": "left",
      "vertical": "bottom",
    }
  }
  animated={true}
  iconButtonElement={
    <span
      style={
        Object {
          "color": "white",
        }
      }>
      Snippets
    </span>
  }
  multiple={false}
  onItemTouchTap={[Function]}
  onKeyboardFocus={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onRequestChange={[Function]}
  onTouchTap={[Function]}
  open={false}
  style={
    Object {
      "alignItems": "center",
      "background": "transparent",
      "borderBottomColor": "transparent",
      "borderBottomWidth": "3px",
      "cursor": "pointer",
      "display": "inline-flex",
      "height": "100%",
      "paddingLeft": "10px",
      "paddingRight": "10px",
    }
  }
  targetOrigin={
    Object {
      "horizontal": "left",
      "vertical": "top",
    }
  }
  touchTapCloseDelay={200}
  useLayerForClickAway={true}>
  <SnippetList
    onClick={[Function]}
    primaryText="My Snippets"
    snippetOwner="foo"
    snippetsList={Object {}} />
  <OrgSnippetsMenu
    onClick={[Function]}
    orgSnippets={Object {}} />
  <MenuItem
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    checked={false}
    desktop={false}
    disabled={false}
    focusState="none"
    insetChildren={false}
    onTouchTap={[Function]}
    primaryText="Search" />
</IconMenu>
`;
